File: frontend/src/types/resource.ts
// 资源类型（存储用）
export const ResourceType = {
  WebMarkdown: 'web_markdown',
  Ppt: 'ppt',
  Pptx: 'pptx',
  Doc: 'doc',
  Docx: 'docx',
  Pdf: 'pdf',
  Txt: 'txt',
  Jpeg: 'jpeg',
  Jpg: 'jpg',
  Png: 'png',
  Zip: 'zip',
  Other: 'other'
} as const;

export type ResourceTypeType = typeof ResourceType[keyof typeof ResourceType];

// 资源类型筛选选项（用于筛选下拉框，合并相关类型）
export const ResourceTypeFilterOptions = {
  ppt: 'ppt',           // PPT（包含 ppt 和 pptx）
  doc: 'doc',           // Word 文档（包含 doc 和 docx）
  pdf: 'pdf',           // PDF
  image: 'image',       // 图片（包含 jpeg, jpg, png）
  web_markdown: 'web_markdown', // 网页 Markdown
  txt: 'txt',           // 文本文件
  zip: 'zip',           // ZIP 压缩包
  other: 'other'        // 其他
} as const;

export type ResourceTypeFilterType = typeof ResourceTypeFilterOptions[keyof typeof ResourceTypeFilterOptions];

// 资源分类
export const ResourceCategory = {
  ExamResult: 'exam_result',
  LearningNote: 'learning_note',
  PastPaper: 'past_paper',
  Note: 'note',
  ReviewOutline: 'review_outline',
  Lecture: 'lecture',
  Other: 'other'
} as const;

export type ResourceCategoryType = typeof ResourceCategory[keyof typeof ResourceCategory];

// 审核状态
export const AuditStatus = {
  Pending: 'pending',
  Approved: 'approved',
  Rejected: 'rejected'
} as const;

export type AuditStatusType = typeof AuditStatus[keyof typeof AuditStatus];

// 资源统计信息
export interface ResourceStats {
  views: number;
  downloads: number;
  likes: number;
  avgDifficulty?: number;
  avgQuality?: number;
  avgDetail?: number;
  ratingCount: number;
}

// 资源列表项
export interface ResourceListItem {
  id: string;
  title: string;
  courseName?: string;
  resourceType: string;
  category: string;
  tags?: string[];
  auditStatus: string;
  createdAt: string;
  stats: ResourceStats;
  uploaderName?: string;
}

// 资源详情
export interface ResourceDetail {
  id: string;
  title: string;
  authorId?: string;
  uploaderId: string;
  courseName?: string;
  resourceType: string;
  category: string;
  tags?: string[];
  description?: string;
  fileSize?: number;
  auditStatus: string;
  createdAt: string;
  updatedAt: string;
  stats: ResourceStats;
  uploaderName?: string;
}

// 资源列表响应
export interface ResourceListResponse {
  resources: ResourceListItem[];
  total: number;
  page: number;
  perPage: number;
}

// 资源列表查询参数
export interface ResourceListQuery {
  page?: number;
  perPage?: number;
  resourceType?: string;
  category?: string;
  sortBy?: 'created_at' | 'downloads' | 'likes' | 'rating' | 'title';
  sortOrder?: 'asc' | 'desc';
}

// 资源搜索查询参数
export interface ResourceSearchQuery {
  q: string;
  page?: number;
  perPage?: number;
  resourceType?: string;
  category?: string;
}

// 上传资源请求
export interface UploadResourceRequest {
  title: string;
  courseName?: string;
  resourceType: ResourceTypeType;
  category: ResourceCategoryType;
  tags?: string[];
  description?: string;
}

// 上传资源响应
export interface UploadResourceResponse {
  id: string;
  title: string;
  resourceType: string;
  auditStatus: string;
  aiMessage?: string;
  createdAt: string;
}

// 资源类型显示名称映射（用于显示单个资源的类型）
export const ResourceTypeLabels: Record<ResourceTypeType, string> = {
  [ResourceType.WebMarkdown]: '网页 Markdown',
  [ResourceType.Ppt]: 'PPT',
  [ResourceType.Pptx]: 'PPTX',
  [ResourceType.Doc]: 'Word 文档',
  [ResourceType.Docx]: 'Word 文档',
  [ResourceType.Pdf]: 'PDF',
  [ResourceType.Txt]: '文本文件',
  [ResourceType.Jpeg]: 'JPEG 图片',
  [ResourceType.Jpg]: 'JPG 图片',
  [ResourceType.Png]: 'PNG 图片',
  [ResourceType.Zip]: 'ZIP 压缩包',
  [ResourceType.Other]: '其他'
};

// 资源类型筛选显示名称映射（用于筛选下拉框，合并相关类型）
export const ResourceTypeFilterLabels: Record<ResourceTypeFilterType, string> = {
  ppt: 'PPT 演示文稿',
  doc: 'Word 文档',
  pdf: 'PDF 文档',
  image: '图片',
  web_markdown: '网页 Markdown',
  txt: '文本文件',
  zip: 'ZIP 压缩包',
  other: '其他'
};

// 资源分类显示名称映射
export const ResourceCategoryLabels: Record<ResourceCategoryType, string> = {
  [ResourceCategory.ExamResult]: '考试成绩分布',
  [ResourceCategory.LearningNote]: '学习心得',
  [ResourceCategory.PastPaper]: '往年试卷',
  [ResourceCategory.Note]: '笔记',
  [ResourceCategory.ReviewOutline]: '复习提纲',
  [ResourceCategory.Lecture]: '讲义',
  [ResourceCategory.Other]: '其他'
};

// 审核状态显示名称映射
export const AuditStatusLabels: Record<AuditStatusType, string> = {
  [AuditStatus.Pending]: '待审核',
  [AuditStatus.Approved]: '已通过',
  [AuditStatus.Rejected]: '已拒绝'
};

// 获取资源类型颜色
export function getResourceTypeColor(type: string): string {
  const colorMap: Record<string, string> = {
    [ResourceType.Pdf]: '#F56C6C',
    [ResourceType.Ppt]: '#E6A23C',
    [ResourceType.Pptx]: '#E6A23C',
    [ResourceType.Doc]: '#409EFF',
    [ResourceType.Docx]: '#409EFF',
    [ResourceType.WebMarkdown]: '#67C23A',
    [ResourceType.Txt]: '#909399',
    [ResourceType.Zip]: '#909399'
  };
  return colorMap[type] || '#909399';
}

// 格式化文件大小
export function formatFileSize(bytes?: number): string {
  if (!bytes) return '-';
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
}

// 支持的文件扩展名
export const SupportedExtensions = [
  'md', 'markdown', 'ppt', 'pptx', 'doc', 'docx',
  'pdf', 'txt', 'jpeg', 'jpg', 'png', 'zip'
];

// 从文件名获取资源类型
export function getResourceTypeFromFileName(fileName: string): ResourceTypeType {
  const ext = fileName.split('.').pop()?.toLowerCase() || '';
  const typeMap: Record<string, ResourceTypeType> = {
    'md': ResourceType.WebMarkdown,
    'markdown': ResourceType.WebMarkdown,
    'ppt': ResourceType.Ppt,
    'pptx': ResourceType.Pptx,
    'doc': ResourceType.Doc,
    'docx': ResourceType.Docx,
    'pdf': ResourceType.Pdf,
    'txt': ResourceType.Txt,
    'jpeg': ResourceType.Jpeg,
    'jpg': ResourceType.Jpg,
    'png': ResourceType.Png,
    'zip': ResourceType.Zip
  };
  return typeMap[ext] || ResourceType.Other;
}

// 更新资源内容请求（用于Markdown在线编辑）
export interface UpdateResourceContentRequest {
  content: string;
}

// 更新资源内容响应
export interface UpdateResourceContentResponse {
  id: string;
  updatedAt: string;
}

// 获取资源原始内容响应
export interface GetResourceRawContentResponse {
  content: string;
}

// 热门资源列表项
export interface HotResourceItem {
  id: string;
  title: string;
  courseName?: string;
  resourceType: string;
  downloads: number;
  views: number;
  likes: number;
}

File: backend/src/models/resource.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

/// 资源类型枚举
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "VARCHAR", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum ResourceType {
    /// 网页 Markdown
    WebMarkdown,
    /// PPT 文件
    Ppt,
    /// PPTX 文件
    Pptx,
    /// Word 文档
    Doc,
    /// Word 文档 (新格式)
    Docx,
    /// PDF 文件
    Pdf,
    /// 文本文件
    Txt,
    /// JPEG 图片
    Jpeg,
    /// JPG 图片
    Jpg,
    /// PNG 图片
    Png,
    /// ZIP 压缩包（源文件）
    Zip,
    /// 其他类型
    Other,
}

impl Default for ResourceType {
    fn default() -> Self {
        ResourceType::Other
    }
}

impl ToString for ResourceType {
    fn to_string(&self) -> String {
        match self {
            ResourceType::WebMarkdown => "web_markdown".to_string(),
            ResourceType::Ppt => "ppt".to_string(),
            ResourceType::Pptx => "pptx".to_string(),
            ResourceType::Doc => "doc".to_string(),
            ResourceType::Docx => "docx".to_string(),
            ResourceType::Pdf => "pdf".to_string(),
            ResourceType::Txt => "txt".to_string(),
            ResourceType::Jpeg => "jpeg".to_string(),
            ResourceType::Jpg => "jpg".to_string(),
            ResourceType::Png => "png".to_string(),
            ResourceType::Zip => "zip".to_string(),
            ResourceType::Other => "other".to_string(),
        }
    }
}

impl ResourceType {
    /// 从文件扩展名推断资源类型
    pub fn from_extension(ext: &str) -> Self {
        match ext.to_lowercase().as_str() {
            "md" | "markdown" => ResourceType::WebMarkdown,
            "ppt" => ResourceType::Ppt,
            "pptx" => ResourceType::Pptx,
            "doc" => ResourceType::Doc,
            "docx" => ResourceType::Docx,
            "pdf" => ResourceType::Pdf,
            "txt" => ResourceType::Txt,
            "jpeg" => ResourceType::Jpeg,
            "jpg" => ResourceType::Jpg,
            "png" => ResourceType::Png,
            "zip" => ResourceType::Zip,
            _ => ResourceType::Other,
        }
    }

    /// 获取支持的文件扩展名列表
    pub fn supported_extensions() -> Vec<&'static str> {
        vec![
            "md", "markdown", "ppt", "pptx", "doc", "docx",
            "pdf", "txt", "jpeg", "jpg", "png", "zip",
        ]
    }

    /// 检查是否支持预览（预留接口）
    #[allow(dead_code)]
    pub fn is_previewable(&self) -> bool {
        matches!(
            self,
            ResourceType::WebMarkdown
                | ResourceType::Pdf
                | ResourceType::Txt
                | ResourceType::Jpeg
                | ResourceType::Jpg
                | ResourceType::Png
        )
    }

    /// 获取 MIME 类型（预留接口）
    #[allow(dead_code)]
    pub fn mime_type(&self) -> &'static str {
        match self {
            ResourceType::WebMarkdown => "text/markdown",
            ResourceType::Ppt => "application/vnd.ms-powerpoint",
            ResourceType::Pptx => "application/vnd.openxmlformats-officedocument.presentationml.presentation",
            ResourceType::Doc => "application/msword",
            ResourceType::Docx => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            ResourceType::Pdf => "application/pdf",
            ResourceType::Txt => "text/plain",
            ResourceType::Jpeg => "image/jpeg",
            ResourceType::Jpg => "image/jpeg",
            ResourceType::Png => "image/png",
            ResourceType::Zip => "application/zip",
            ResourceType::Other => "application/octet-stream",
        }
    }
}

/// 资源分类枚举
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "VARCHAR", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum ResourceCategory {
    /// 考试成绩分布
    ExamResult,
    /// 学习心得
    LearningNote,
    /// 往年试卷
    PastPaper,
    /// 笔记
    Note,
    /// 复习提纲
    ReviewOutline,
    /// 讲义
    Lecture,
    /// 其他
    Other,
}

impl Default for ResourceCategory {
    fn default() -> Self {
        ResourceCategory::Other
    }
}

impl ToString for ResourceCategory {
    fn to_string(&self) -> String {
        match self {
            ResourceCategory::ExamResult => "exam_result".to_string(),
            ResourceCategory::LearningNote => "learning_note".to_string(),
            ResourceCategory::PastPaper => "past_paper".to_string(),
            ResourceCategory::Note => "note".to_string(),
            ResourceCategory::ReviewOutline => "review_outline".to_string(),
            ResourceCategory::Lecture => "lecture".to_string(),
            ResourceCategory::Other => "other".to_string(),
        }
    }
}

/// 审核状态枚举
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "VARCHAR", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum AuditStatus {
    /// 待审核
    Pending,
    /// 已通过
    Approved,
    /// 已拒绝
    Rejected,
}

impl Default for AuditStatus {
    fn default() -> Self {
        AuditStatus::Pending
    }
}

impl ToString for AuditStatus {
    fn to_string(&self) -> String {
        match self {
            AuditStatus::Pending => "pending".to_string(),
            AuditStatus::Approved => "approved".to_string(),
            AuditStatus::Rejected => "rejected".to_string(),
        }
    }
}

/// 资源结构体（对应数据库 resources 表）
#[derive(Debug, Clone, FromRow, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Resource {
    pub id: Uuid,
    pub title: String,
    pub author_id: Option<Uuid>,
    pub uploader_id: Uuid,
    pub course_name: Option<String>,
    pub resource_type: String,
    pub category: String,
    pub tags: Option<serde_json::Value>,
    pub file_path: String,
    pub source_file_path: Option<String>,
    pub file_hash: Option<String>,
    pub file_size: Option<i64>,
    pub content_accuracy: Option<f64>,
    pub audit_status: String,
    pub ai_reject_reason: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

/// 资源统计信息（对应数据库 resource_stats 表）
#[derive(Debug, Clone, FromRow, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceStats {
    pub resource_id: Uuid,
    pub views: i32,
    pub downloads: i32,
    pub likes: i32,
    pub avg_difficulty: Option<f64>,
    pub avg_quality: Option<f64>,
    pub avg_detail: Option<f64>,
    pub rating_count: i32,
}

/// 资源上传请求 DTO
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadResourceRequest {
    pub title: String,
    pub course_name: Option<String>,
    // 前端传入的资源类型，后端实际从文件扩展名推断（保留用于API兼容性）
    #[allow(dead_code)]
    pub resource_type: ResourceType,
    pub category: ResourceCategory,
    pub tags: Option<Vec<String>>,
    pub description: Option<String>,
}

impl UploadResourceRequest {
    /// 验证上传请求
    pub fn validate(&self) -> Result<(), String> {
        // 标题验证
        if self.title.trim().is_empty() {
            return Err("资源标题不能为空".to_string());
        }
        if self.title.len() > 255 {
            return Err("资源标题不能超过255个字符".to_string());
        }

        // 标签验证（如果提供）
        if let Some(tags) = &self.tags {
            if tags.len() > 10 {
                return Err("标签数量不能超过10个".to_string());
            }
            for tag in tags {
                if tag.len() > 50 {
                    return Err("单个标签不能超过50个字符".to_string());
                }
            }
        }

        Ok(())
    }
}

/// 资源上传响应 DTO
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadResourceResponse {
    pub id: Uuid,
    pub title: String,
    pub resource_type: String,
    pub audit_status: String,
    pub ai_message: Option<String>,
    pub created_at: NaiveDateTime,
}

/// 资源详情响应 DTO
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceDetailResponse {
    pub id: Uuid,
    pub title: String,
    pub author_id: Option<Uuid>,
    pub uploader_id: Uuid,
    pub course_name: Option<String>,
    pub resource_type: String,
    pub category: String,
    pub tags: Option<Vec<String>>,
    pub description: Option<String>,
    pub file_size: Option<i64>,
    pub audit_status: String,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    pub stats: ResourceStatsResponse,
    pub uploader_name: Option<String>,
}

/// 资源统计响应 DTO
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceStatsResponse {
    pub views: i32,
    pub downloads: i32,
    pub likes: i32,
    pub avg_difficulty: Option<f64>,
    pub avg_quality: Option<f64>,
    pub avg_detail: Option<f64>,
    pub rating_count: i32,
}

/// 资源列表响应 DTO
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceListResponse {
    pub resources: Vec<ResourceListItem>,
    pub total: i64,
    pub page: i32,
    pub per_page: i32,
}

/// 资源列表项 DTO
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceListItem {
    pub id: Uuid,
    pub title: String,
    pub course_name: Option<String>,
    pub resource_type: String,
    pub category: String,
    pub tags: Option<Vec<String>>,
    pub audit_status: String,
    pub created_at: NaiveDateTime,
    pub stats: ResourceStatsResponse,
    pub uploader_name: Option<String>,
}

/// 资源列表查询参数
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceListQuery {
    pub page: Option<i32>,
    pub per_page: Option<i32>,
    pub resource_type: Option<String>,
    pub category: Option<String>,
    pub sort_by: Option<String>,
    pub sort_order: Option<String>,
}

/// 资源搜索查询参数
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceSearchQuery {
    pub q: String,
    pub page: Option<i32>,
    pub per_page: Option<i32>,
    pub resource_type: Option<String>,
    pub category: Option<String>,
}

impl ResourceListQuery {
    pub fn get_page(&self) -> i32 {
        self.page.unwrap_or(1).max(1)
    }

    pub fn get_per_page(&self) -> i32 {
        self.per_page.unwrap_or(20).min(100).max(1)
    }
}

impl ResourceSearchQuery {
    pub fn get_page(&self) -> i32 {
        self.page.unwrap_or(1).max(1)
    }

    pub fn get_per_page(&self) -> i32 {
        self.per_page.unwrap_or(20).min(100).max(1)
    }
}

/// AI 审核结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiAuditResult {
    pub passed: bool,
    pub reason: Option<String>,
    pub accuracy_score: Option<f64>,
}

/// 更新资源内容请求 DTO（用于Markdown在线编辑）
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateResourceContentRequest {
    pub content: String,
}

impl UpdateResourceContentRequest {
    /// 验证请求
    pub fn validate(&self) -> Result<(), String> {
        // 内容不能为空
        if self.content.trim().is_empty() {
            return Err("内容不能为空".to_string());
        }
        // 内容长度限制（10MB）
        if self.content.len() > 10 * 1024 * 1024 {
            return Err("内容大小超过10MB限制".to_string());
        }
        Ok(())
    }
}

/// 更新资源内容响应 DTO
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateResourceContentResponse {
    pub id: Uuid,
    pub updated_at: chrono::NaiveDateTime,
}

/// 热门资源查询参数
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HotResourcesQuery {
    pub limit: Option<i32>,
}

/// 热门资源列表项 DTO
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HotResourceItem {
    pub id: Uuid,
    pub title: String,
    pub course_name: Option<String>,
    pub resource_type: String,
    pub downloads: i32,
    pub views: i32,
    pub likes: i32,
}
